README
This is the ML version of the New Jersey Machine-Code Toolkit.
It was prepared on Mon Aug  7 11:32:46 EDT 2000.
For information about what to do with it, please consult
http://www.eecs.harvard.edu/~nr/toolkit/ml.html.


src/CHANGES
For ???

48. Simplify slice in case for Ewiden in M3 emitter. 11/12/96

47. Changed the way the toolkit generates matchers, so there are no
    variables declared to hold absolute fields.  Such variables were
    dead in many cases.  As a result, added Gcommented to comment such
    expressions, and modified the simplifier generator to insert
    Gcommented in all rules (where needed).  11/12/96

46. For early branches, merge fits conditions in with the pattern
    conditions, so they can be used to distinguish disjuncts.  Fixes
    bug 15 and cleans up the sparc set constructor.  11/8/96

45. Use Eapp in grammar, but fix [[constructors_applied_in]] to handle
    [[Eapp]].  10/23/96

44. Warn when we don't see what's expected in fetch templates (e.g.,
    fetch without %a).  10/23/96

43. When generating code for a single node in decision tree, make sure
    to have declarations for fields and matching words used in
    conditions. Fixes #18.  10/23/96 

42. Now eliminate instances in a more elaborate way, which preserves
    info about nested instances long enough to do better
    elimination. Fixes #17.  10/21/96

41. Make sure initial ( goes into call to print proc even if
    constructor has no operands. 8/13/96

40. Specify one-bit fields with one integer, by analogy with
    one-integer specifications of one-bit slices. May 96

39. Reverse bit numbering a total botch because tried to combine
    negate and swap.  Fixed May 96.

38. Include pattern length in dag info.  Fixes #14.  6/7/96

37. In solve, substitute new values in constraints as well as other
    places.  Failure to do this tickled a bug in building a decoder
    with conditions like {n != 0} where n had to be extracted from an
    absolute field. 6/7/96

36. Added code to simplify  n[!K] = n@[0:K-1] to FITSS(n, K), but this
    introduced a bug in decoding, so made it part of super_simplify
    only.  The original improves condition checking in encoding
    functions for the bytecode interpreter. 6/7/96

35. Changed code in Cexp to generate an empty statement to emit epsilon. 4/10/96

34. Emit one header per closure function, not one per encoding function.

For version 0.4:

33. When no word matches a pattern, toolkit now finds covering patterns. 3/3/96

32. Can now use _ as a wildcard in any expression --- gets turned into a fresh
    variable in the grammar. 3/3/96

31. Sign-extended variables now get declared int in matching statements 3/3/96

30. stands for myriad changes that are too big to record

For version 0.2:

29. Change the signed and unsigned arg parts of constructor application so
    that field names are always treated as free variables.  This lets us use
    the name of a field as a binding instance in a matching statement.  God
    only knows what other consequences it may have. 9/1/95

28. Use single if statement for some nodes in decoders. 8/18/95

27. Support for alternate bit numbering.  7/19/95

26. Support for octal and binary literals. 7/19/95

25. Many unlogged changes, including using expressions as constructor
    args and in field bindings.

For version 0.1b:

24. Mysterious & in balance.nw was source of need for *ws?!  No bugs in icont!
    11/6/94

For version 0.1a:

23. Change printing of linear sums to print out positive terms, then
    negative terms, with constant term always last in its group. 11/3/94

22. Simplify constraints in solver for prettier output. 11/2/94

21. Pursuant to bug #9, aalwaysmatches() in apply() in tree.nw is too
    optimistic, so added call to guard_always_satisfied in apply(). 11/1/94

20. Copy case statement in tree(), and pass original case statment to
    checktree() to fix #12 (unmatching arms silently lost). 11/1/94 

19. Initialize root node with empty children in tree(). 11/1/94

18. Changed rejection test for narrow/widen balances in balance.nw
    Old test rejected balance a! = b when we had already seen a! = c,
    but then we lost the information that b = c, leading to bug 8. 11/1/94

17. More debugging for the balancer, -debug-bal option. 11/1/94

16. Move narrow(widen(x)) -> x from super to regular rewrite rules. 11/1/94

15. Error message in match.nw shows just the identifiers not solved for. 11/1/94

14. Moved call of Generate_C() in main.nw to precede argument parsing. 10/31/94

13. Added synthetic branches, multiply, load and move float to mips
    specification (removed nop() calls from synthetics). 10/25/94

12. Generate_M3() had an old bogus call.  Removed.  10/20/94.

For version 0.1:

11. Change simplify to call itself using simfun macro, and
    created super_simplify. 10/17/94

10. Removed dclproto from mclib.nw.  Cleaned up chunk names for
    mclib.h. 10/17/94

 9. Added definition of sign_extend to asm-header.t in assembly.nw,
    because it was needed to implement Ewiden. 10/17/94

 8. Changed representation of groups of tokens for emission.  Now
    gather them all in single list, and store the offset from the
    beginning of each token.  This simplifies correct handling of $pc
    per bugs #4 and #5.  Changes to exp, expwalk, constructors, and
    Cexp. 10/17/94

 7. Eliminated save_closure, renamed alloc_closure to mc_alloc_closure,
    pulled alloc_closure from mclib.nw, and eliminated RClosure_pool.
    Didn't want two procedures always called one after the other.
    Explanation in library.nw. 10/17/94

 6. Changed buffer_at_lc to point to preceding buffer when lc falls on
    boundary between buffers (see bug #3). 10/17/94

 5. Beefed up simplification of disjuncts and adisjuncts to eliminate
    disjuncts containing unfulfillable range constraints. 10/2/94.

 4. Eliminated some rewrite rules as too strength-reducing, making it
    possible to use the simplifier in match.nw, so that we could
    eliminate unsolvable disjuncts that didn't make sense (see bug #2).
    10/3/94

 3. Removed declaration of sign_extend from assembly.nw; it is provided
    by Cexp.nw when needed.


src/tests/sret.m
match [nextPC] lc to
| ret() =>
        lc = nextPC;
        return true;
else
        return false;
endmatch


base/mclib.h
#ifndef INCLUDE_MCLIBH
#define INCLUDE_MCLIBH

#ifndef assert
#include <assert.h>
#endif 

/* this group is available for users */

typedef struct reloc_block *RBlock;
typedef struct label *RLabel;
typedef struct rclosure *RClosure;
typedef void (*Emitter) 
    (RBlock, unsigned dest_lc, unsigned long bits, unsigned size);
typedef void (*FailCont)(char *fmt, ...);
typedef enum mc_alloc_pool { 
    RBlock_pool=1, RBlock_data_pool, RLabel_pool, RAddr_pool
} Mc_alloc_pool;
typedef struct relocatable_address *RAddr;
extern RBlock block_new(unsigned size);
extern void set_lc(RBlock rb, unsigned lc);
extern void set_block(RBlock rb);
extern void set_address(RBlock, unsigned);
extern void set_register(RBlock rb, unsigned reg, unsigned regaddr);
extern unsigned block_reg(RBlock rb);
extern unsigned block_regaddr(RBlock rb);
extern RLabel label_new(char *name);
extern void label_define(RLabel lbl, int offset);
extern void label_define_at(RLabel lbl, RBlock block, unsigned lc);
extern void (*emitm)(unsigned long value, unsigned n);
extern void emitl(unsigned long value, unsigned n);
extern void emitb(unsigned long value, unsigned n);
extern void align(unsigned n);
extern void apply_closure(RClosure cl, Emitter emitter, FailCont failc);
extern void (*cl_emitm)(RBlock dest_block, unsigned dest_lc, 
                           unsigned long bits, unsigned size);
extern void cl_emitl (RBlock dest_block, unsigned dest_lc, 
                           unsigned long bits, unsigned size);
extern void cl_emitb (RBlock dest_block, unsigned dest_lc, 
                           unsigned long bits, unsigned size);
extern void block_write(int fd,RBlock rb,unsigned start,unsigned n);
extern void block_copy(unsigned char *buf,
                       RBlock rb, unsigned start, unsigned n);
extern unsigned long (*block_fetchm)(RBlock dest_block, 
                                     unsigned dest_lc, unsigned size);
extern unsigned long   block_fetchl (RBlock dest_block, 
                                     unsigned dest_lc, unsigned size);
extern unsigned long   block_fetchb (RBlock dest_block, 
                                     unsigned dest_lc, unsigned size);
extern void (*asmprintf)(void *closure, const char *fmt, ...);
extern void *asmprintfd; /* closure */
extern void (*asmprintreloc)(RAddr reloc); /* calls asmprintf */
extern RAddr addr_new(RLabel label, int offset);
extern RAddr unsigned_to_raddr(unsigned lc);
extern void *mc_alloc(int size, Mc_alloc_pool pool);
extern RClosure mc_alloc_closure(unsigned size, 
                         RBlock dest_block, unsigned dest_lc);

/* this group either supports macros or is needed by emission procedures */

extern void (*fail)(char *fmt, ...);
struct relocatable_address {
    struct label *label;
    int offset;
};
typedef struct rblock_buffer {
  unsigned char *data, *limit;
  struct rblock_buffer *next;
  unsigned datalc;
} *Buffer;
struct reloc_block { 
  struct rblock_buffer firstbuf, *curbuf;
  unsigned char *p;     /* points to current lc -- dirty */
  unsigned size;        /* highest lc ever -- dirty */
  struct { unsigned address:1, reg:1; } known;
  unsigned address;       /* absolute address of block in memory */
  unsigned reg;           /* number of register pointing into this block */
  unsigned reglc;         /* when reg non-null, location to which it points */
  struct label *label;    /* (anonymous) label associated with this block */
  int size_hint;
  struct reloc_block *next;       /* instrumentation */
};
#define crb() currb
extern RBlock currb;              /* the current relocatable block */
extern unsigned char *currb_p;    /* the true version of currb->p */
extern unsigned char *currb_safe; 
              /* the true version of currb->curbuf->limit - MARGIN */
extern RBlock defined_rbs;      /* instrumentation */
extern int nblocks;             /* instrumentation */
struct label {
    RBlock block;
    unsigned offset;
    char *name;
};
typedef void (*RelocCallback)(void *closure, RAddr reloc);
typedef void (*RelocEnumerator)(struct rclosure *cp, RelocCallback f, void *closure);

typedef struct instance {
    int tag;         /* need at least 11 bits for Intel */
    union { struct { void *inputs[1]; } consname; } u;
} Instance;
typedef void (*ApplyMethod)(RClosure cl, Emitter emitter, FailCont fail);

typedef struct closure_header {
    ApplyMethod apply;
    RelocEnumerator relocfn;
    int uses_pc;
    int size_in_bytes;
} *ClosureHeader;       /* always allocated statically */

typedef struct closure_location {
    RBlock dest_block;        /* block written */
    unsigned dest_lc;         /* location within block to write */
} ClosureLocation;

struct rclosure {
    ClosureHeader h;
    ClosureLocation loc;
    /* fields may follow */
};
typedef struct closurepostfix {
    ApplyMethod apply;
    char *postfix;
} ClosurePostfix;
#define location_known(raddr)      ((raddr) && label_location_known((raddr)->label))
#define location(raddr)            (label_location((raddr)->label) + (raddr)->offset)
#define MARGIN 8
#define emitul(BITS) emitm((BITS), sizeof(unsigned long))
#define emitu(BITS)  emitm((BITS), sizeof(unsigned))
#define emitus(BITS) emitm((BITS), sizeof(unsigned short))
#define emituc(BITS) emitm((BITS), sizeof(unsigned char))
#define emitfastT0(BITS, TYPE) do { \
  if (currb_p < currb_safe) { \
     *((TYPE*)currb_p) = (TYPE) (BITS); \
      currb_p += sizeof(TYPE); \
  } else (*emitm)((BITS), sizeof(TYPE)); \
} while (0)
#define emitfastul0(BITS) emitfastT0((BITS), unsigned long)
#define emitfastu0(BITS)  emitfastT0((BITS), unsigned)
#define emitfastus0(BITS) emitfastT0((BITS), unsigned short)
#define emitfastuc0(BITS) emitfastT0((BITS), unsigned char)
#define emitfast0(BITS, SIZE) do { \
  if      ((SIZE) == sizeof(unsigned long))  emitfastul0((BITS)); \
  else if ((SIZE) == sizeof(unsigned))       emitfastu0((BITS)); \
  else if ((SIZE) == sizeof(unsigned short)) emitfastus0((BITS)); \
  else if ((SIZE) == sizeof(unsigned char))  emitfastuc0((BITS)); \
  else emitm((BITS), (SIZE)); \
} while(0)
#define emitfastbyte0(BITS, SIZE) do { \
  if ((SIZE) == sizeof(unsigned char)) \
    emitfastuc0((BITS)); \
  else \
    emitm((BITS), (SIZE)); \
} while(0)
#define emitfastT1(BITS, TYPE) do { \
  register unsigned char *p = currb_p; \
  if (p < currb_safe) { \
     *((TYPE*)p) = (TYPE) (BITS); \
      currb_p = p + sizeof(TYPE); \
  } else (*emitm)((BITS), sizeof(TYPE)); \
} while (0)
#define emitfastul1(BITS) emitfastT1((BITS), unsigned long)
#define emitfastu1(BITS)  emitfastT1((BITS), unsigned)
#define emitfastus1(BITS) emitfastT1((BITS), unsigned short)
#define emitfastuc1(BITS) emitfastT1((BITS), unsigned char)
#define emitfast1(BITS, SIZE) do { \
  if      ((SIZE) == sizeof(unsigned long))  emitfastul1((BITS)); \
  else if ((SIZE) == sizeof(unsigned))       emitfastu1((BITS)); \
  else if ((SIZE) == sizeof(unsigned short)) emitfastus1((BITS)); \
  else if ((SIZE) == sizeof(unsigned char))  emitfastuc1((BITS)); \
  else emitm((BITS), (SIZE)); \
} while(0)
#define emitfastbyte1(BITS, SIZE) do { \
  if ((SIZE) == sizeof(unsigned char)) \
    emitfastuc1((BITS)); \
  else \
    emitm((BITS), (SIZE)); \
} while(0)
#define emitfastT2(BITS, TYPE) ( \
  (EMITFAST_p < currb_safe) \
    ? (*((TYPE*)EMITFAST_p) = (TYPE) (BITS), \
       currb_p = EMITFAST_p + sizeof(TYPE), \
       (void)0) \
    : (*emitm)((BITS), sizeof(TYPE)) \
  )
#define Xemitfastul2(BITS) emitfastT2((BITS), unsigned long)
#define Xemitfastu2(BITS)  emitfastT2((BITS), unsigned)
#define Xemitfastus2(BITS) emitfastT2((BITS), unsigned short)
#define Xemitfastuc2(BITS) emitfastT2((BITS), unsigned char)
#define emitfast2(BITS, SIZE) do { \
  register unsigned char *EMITFAST_p = currb_p; \
  ((SIZE) == sizeof(unsigned long))  ? Xemitfastul2((BITS)) : \
  ((SIZE) == sizeof(unsigned))       ? Xemitfastu2((BITS))  : \
  ((SIZE) == sizeof(unsigned short)) ? Xemitfastus2((BITS)) : \
  ((SIZE) == sizeof(unsigned char))  ? Xemitfastuc2((BITS)) : \
  emitm((BITS), (SIZE)); \
} while(0)
#define emitfastbyte2(BITS, SIZE) do { \
  register unsigned char *EMITFAST_p = currb_p; \
  ((SIZE) == sizeof(unsigned char)) ? emitfastuc2((BITS)) : \
  emitm((BITS), (SIZE)); \
} while(0)
#define emitfastul2(BITS) do { \
  register unsigned char *EMITFAST_p = currb_p; Xemitfastul2((BITS)); \
} while(0)
#define emitfastu2(BITS) do { \
  register unsigned char *EMITFAST_p = currb_p; Xemitfastu2((BITS)); \
} while(0)
#define emitfastus2(BITS) do { \
  register unsigned char *EMITFAST_p = currb_p; Xemitfastus2((BITS)); \
} while(0)
#define emitfastuc2(BITS) do { \
  register unsigned char *EMITFAST_p = currb_p; Xemitfastuc2((BITS)); \
} while(0)
#define emitfast        emitfast2
#define emitfastbyte    emitfastbyte2
#define emitfastul      emitfastul2
#define emitfastu       emitfastu2
#define emitfastus      emitfastus2
#define emitfastuc      emitfastuc2
#define lc() (currb_p - currb->curbuf->data + currb->curbuf->datalc)
#define org(LOC)        set_lc(currb,(LOC))
#define addlc(N) ((void)(currb_p += (N)))
#define block_lc(rb) \
  ((rb) == currb \
      ? (make_clean(),((rb)->p - (rb)->curbuf->data + (rb)->curbuf->datalc)) \
      :               ((rb)->p - (rb)->curbuf->data + (rb)->curbuf->datalc))
#define block_low(rb) \
  ((rb) == currb \
      ? (make_clean(), (rb)->firstbuf.datalc) \
      :                (rb)->firstbuf.datalc)
extern struct reloc_block undefrblock;
#define block_defined(rb) ((rb) && (rb) != &undefrblock)
#define block_label(rb) ((rb)->label)
#define block_address_known(rb) ((rb)->known.address)
#define block_address(rb)       ((rb)->address)
#define block_size(rb) ((rb) == currb ? (make_clean(), (rb)->size) : (rb)->size)
#define label_location_known(lbl)  (((lbl) && (lbl)->block->known.address))
#define label_location(lbl)        ((lbl)->block->address + (lbl)->offset)
#define pc_location_known(loc)     ((loc).dest_block->known.address)
#define pc_location(loc)           ((loc).dest_block->address + (loc).dest_lc)
#define cur_pc_known()             (currb->known.address)
#define cur_pc()                   (currb->address + lc())
void make_clean(void);
extern RClosure mc_create_closure_at_offset(unsigned size, ClosureHeader h, unsigned offset);
extern RClosure mc_create_closure_here(unsigned size, ClosureHeader h);

#endif


examples/xs/xs.m.d
#define sign_extend(N,SIZE) (((int)((N) << (sizeof(unsigned)*8-(SIZE)))) >> (sizeof(unsigned)*8-(SIZE)))
#include <assert.h>

#line 2 "../../../toolkit/www/examples/xs/xs.m"
#line 243 "xs.nw"
#include <stdio.h>
#include <stdlib.h>
#line 36 "xs.nw"
typedef struct machine {
  unsigned MEM[1024];
  unsigned AC, XR, PC;
  unsigned PCHI:1;
} *Machine;
#line 48 "xs.nw"
#define validate(A) ((A) & 0x3ff)
#line 72 "xs.nw"
#define setPC(M, A)  (((M)->PC = validate(A)), ((M)->PCHI = 0))
#define advancePC(M) ((M)->PCHI ? setPC(M, (M)->PC+1) : ((M)->PCHI = 1))
#define getPC(A)     ((A)->PC)
#line 173 "xs.nw"
#define FETCH16(A) ((A)->PCHI ? (A)->MEM[(A)->PC] >> 16 : (A)->MEM[(A)->PC] & 0xffff)
#line 87 "xs.nw"
unsigned address(Machine M) {


#line 20 "../../../toolkit/www/examples/xs/xs.m"
{ 
  Machine MATCH_p = 
    
    #line 20 "../../../toolkit/www/examples/xs/xs.m"
    M
    ;
  unsigned /* [0..65535] */ MATCH_w_16_0;
  { 
    MATCH_w_16_0 = FETCH16(MATCH_p); 
    
      switch((MATCH_w_16_0 >> 10 & 0x3) /* md at 0 */) {
        case 0: 
          { 
            unsigned adr = (MATCH_w_16_0 & 0x3ff) /* adr at 0 */;
            
            #line 21 "../../../toolkit/www/examples/xs/xs.m"
             return adr;

            
            
            
          }
          
          break;
        case 1: 
          { 
            unsigned adr = (MATCH_w_16_0 & 0x3ff) /* adr at 0 */;
            
            #line 22 "../../../toolkit/www/examples/xs/xs.m"
             return adr + M->XR;

            
            
            
          }
          
          break;
        case 2: 
          { 
            unsigned adr = (MATCH_w_16_0 & 0x3ff) /* adr at 0 */;
            
            #line 23 "../../../toolkit/www/examples/xs/xs.m"
             return adr + M->XR++;

            
            
            
          }
          
          break;
        case 3: 
          { 
            unsigned adr = (MATCH_w_16_0 & 0x3ff) /* adr at 0 */;
            
            #line 24 "../../../toolkit/www/examples/xs/xs.m"
             return adr + --M->XR;

            
            
            
          }
          
          break;
        default: assert(0);
      } /* (MATCH_w_16_0 >> 10 & 0x3) -- md at 0 --*/ 
    
  }goto MATCH_finished_h; 
  
  MATCH_finished_h: (void)0; /*placeholder for label*/
  
}

#line 27 "../../../toolkit/www/examples/xs/xs.m"
}
#line 121 "xs.nw"
static void show_inst(Machine m);

void interpretXS(Machine m, unsigned start_address) {
  unsigned trace = 0;
  unsigned branch;
#define MEMORY(a) (m->MEM[validate(address(a))])
#define BRANCH(a) (setPC(m, validate(address(a))), branch = 1)
  for (setPC(m, validate(start_address)); ; branch ? 0 : advancePC(m)) {
    if (trace) show_inst(m);
    branch = 0;


#line 38 "../../../toolkit/www/examples/xs/xs.m"
{ 
  Machine MATCH_p = 
    
    #line 38 "../../../toolkit/www/examples/xs/xs.m"
    m
    ;
  unsigned /* [0..65535] */ MATCH_w_16_0;
  { 
    MATCH_w_16_0 = FETCH16(MATCH_p); 
    
      switch((MATCH_w_16_0 >> 12 & 0xf) /* op at 0 */) {
        case 0: 
          if (11 <= (MATCH_w_16_0 & 0x3ff) /* adr at 0 */ && 
            (MATCH_w_16_0 & 0x3ff) /* adr at 0 */ < 1024) 
            
            #line 50 "../../../toolkit/www/examples/xs/xs.m"
             assert(("invalid instruction", 0));

            
             /*opt-block+*/
          else 
            switch((MATCH_w_16_0 & 0x3ff) /* adr at 0 */) {
              case 0: 
                
                #line 39 "../../../toolkit/www/examples/xs/xs.m"
                 return;

                
                
                
                break;
              case 1: 
                
                #line 40 "../../../toolkit/www/examples/xs/xs.m"
                 m->AC = - m->AC;

                
                
                
                break;
              case 2: 
                
                #line 41 "../../../toolkit/www/examples/xs/xs.m"
                 m->AC = ~ m->AC;

                
                
                
                break;
              case 3: 
                
                #line 42 "../../../toolkit/www/examples/xs/xs.m"
                 m->AC <<= 1;

                
                
                
                break;
              case 4: 
                
                #line 43 "../../../toolkit/www/examples/xs/xs.m"
                 m->AC >>= 1;

                
                
                
                break;
              case 5: 
                
                #line 44 "../../../toolkit/www/examples/xs/xs.m"
                 m->AC = getchar();

                
                
                
                break;
              case 6: 
                
                #line 45 "../../../toolkit/www/examples/xs/xs.m"
                 putchar(m->AC);

                
                
                
                break;
              case 7: 
                
                #line 46 "../../../toolkit/www/examples/xs/xs.m"
                 putchar('\n');

                
                
                
                break;
              case 8: 
                
                #line 47 "../../../toolkit/www/examples/xs/xs.m"
                 /* do nothing */;

                
                
                
                break;
              case 9: 
                
                #line 48 "../../../toolkit/www/examples/xs/xs.m"
                 trace = 1;

                
                
                
                break;
              case 10: 
                
                #line 49 "../../../toolkit/www/examples/xs/xs.m"
                 trace = 0;

                
                
                
                break;
              default: assert(0);
            } /* (MATCH_w_16_0 & 0x3ff) -- adr at 0 --*/ 
          break;
        case 1: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 51 "../../../toolkit/www/examples/xs/xs.m"
             m->AC = address(m);

            
            
            
          }
          
          break;
        case 2: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 52 "../../../toolkit/www/examples/xs/xs.m"
             m->AC = MEMORY(m);

            
            
            
          }
          
          break;
        case 3: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 53 "../../../toolkit/www/examples/xs/xs.m"
             MEMORY(m) = m->AC;

            
            
            
          }
          
          break;
        case 4: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 54 "../../../toolkit/www/examples/xs/xs.m"
             m->XR = MEMORY(m);

            
            
            
          }
          
          break;
        case 5: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 55 "../../../toolkit/www/examples/xs/xs.m"
             MEMORY(m) = m->XR;

            
            
            
          }
          
          break;
        case 6: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 56 "../../../toolkit/www/examples/xs/xs.m"
             m->AC += MEMORY(m);

            
            
            
          }
          
          break;
        case 7: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 57 "../../../toolkit/www/examples/xs/xs.m"
             m->AC -= MEMORY(m);

            
            
            
          }
          
          break;
        case 8: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 58 "../../../toolkit/www/examples/xs/xs.m"
             m->AC |= MEMORY(m);

            
            
            
          }
          
          break;
        case 9: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 59 "../../../toolkit/www/examples/xs/xs.m"
             m->AC &= MEMORY(m);

            
            
            
          }
          
          break;
        case 10: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 60 "../../../toolkit/www/examples/xs/xs.m"
             MEMORY(m)++;

            
            
            
          }
          
          break;
        case 11: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 61 "../../../toolkit/www/examples/xs/xs.m"
             MEMORY(m)--;

            
            
            
          }
          
          break;
        case 12: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 62 "../../../toolkit/www/examples/xs/xs.m"
             BRANCH(m);

            
            
            
          }
          
          break;
        case 13: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 63 "../../../toolkit/www/examples/xs/xs.m"
             if (m->AC == 0) then BRANCH(m);

            
            
            
          }
          
          break;
        case 14: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 64 "../../../toolkit/www/examples/xs/xs.m"
             if (m->AC <  0) then BRANCH(m);

            
            
            
          }
          
          break;
        case 15: 
          { 
            unsigned a = getPC(MATCH_p);
            
            #line 65 "../../../toolkit/www/examples/xs/xs.m"
             m->XR = m->PC; BRANCH(m);

            
            
            
          }
          
          break;
        default: assert(0);
      } /* (MATCH_w_16_0 >> 12 & 0xf) -- op at 0 --*/ 
    
  }goto MATCH_finished_g; 
  
  MATCH_finished_g: (void)0; /*placeholder for label*/
  
}

#line 68 "../../../toolkit/www/examples/xs/xs.m"
  }
}
#line 177 "xs.nw"
static char *adrstring(Machine m);

static void show_inst(Machine m) {
  fprintf(stderr, "%03x%s : ", m->PC, m->PCHI ? "'" : " ");


#line 74 "../../../toolkit/www/examples/xs/xs.m"
{ 
  Machine MATCH_p = 
    
    #line 74 "../../../toolkit/www/examples/xs/xs.m"
    m
    ;
  char *MATCH_name;
  static char *MATCH_name_adr_0[] = {
    "HALT", "NEG", "COM", "SHL", "SHR", "READ", "WRT", "NEWL", "NOOP", "TRA", 
    "NOTR", 
  };
  static char *MATCH_name_op_1[] = {
    (char *)0, "LIT", "LDA", "STA", "LDX", "STX", "ADD", "SUB", "OR", "AND", 
    "INC", "DEC", "JMP", "JPZ", "JPN", "JSR", 
  };
  unsigned /* [0..65535] */ MATCH_w_16_0;
  { 
    MATCH_w_16_0 = FETCH16(MATCH_p); 
    if ((MATCH_w_16_0 >> 12 & 0xf) /* op at 0 */ == 0) 
      if (11 <= (MATCH_w_16_0 & 0x3ff) /* adr at 0 */ && 
        (MATCH_w_16_0 & 0x3ff) /* adr at 0 */ < 1024) 
        
        #line 76 "../../../toolkit/www/examples/xs/xs.m"
         fprintf(stderr, "INVALID [%08x]", m->MEM[validate(m->PC)]);

        
         /*opt-block+*/
      else { 
        MATCH_name = MATCH_name_adr_0[(MATCH_w_16_0 & 0x3ff) /* adr at 0 */]; 
        { 
          char *name = MATCH_name;
          
          #line 75 "../../../toolkit/www/examples/xs/xs.m"
           fprintf(stderr, "%s", name);

          
          
          
        }
        
      } /*opt-block*/ /*opt-block+*/
    else { 
      MATCH_name = MATCH_name_op_1[(MATCH_w_16_0 >> 12 & 0xf) /* op at 0 */]; 
      { 
        char *name = MATCH_name;
        unsigned a = getPC(MATCH_p);
        
        #line 77 "../../../toolkit/www/examples/xs/xs.m"
         fprintf(stderr, "%s %s", name, adrstring(m));

        
        
        
      }
      
    } /*opt-block*/
    
  }goto MATCH_finished_f; 
  
  MATCH_finished_f: (void)0; /*placeholder for label*/
  
}

#line 80 "../../../toolkit/www/examples/xs/xs.m"
  fprintf(stderr, "  (AC=%08x, XR=%08x)\n", m->AC, m->XR);
}

static char *adrstring(Machine m) {
  static char buf[100];


#line 84 "../../../toolkit/www/examples/xs/xs.m"
{ 
  Machine MATCH_p = 
    
    #line 84 "../../../toolkit/www/examples/xs/xs.m"
    m
    ;
  unsigned /* [0..65535] */ MATCH_w_16_0;
  { 
    MATCH_w_16_0 = FETCH16(MATCH_p); 
    
      switch((MATCH_w_16_0 >> 10 & 0x3) /* md at 0 */) {
        case 0: 
          { 
            unsigned adr = (MATCH_w_16_0 & 0x3ff) /* adr at 0 */;
            
            #line 85 "../../../toolkit/www/examples/xs/xs.m"
             sprintf(buf, "%08x", adr);

            
            
            
          }
          
          break;
        case 1: 
          { 
            unsigned adr = (MATCH_w_16_0 & 0x3ff) /* adr at 0 */;
            
            #line 86 "../../../toolkit/www/examples/xs/xs.m"
             sprintf(buf, "XR#%08x", adr);

            
            
            
          }
          
          break;
        case 2: 
          { 
            unsigned adr = (MATCH_w_16_0 & 0x3ff) /* adr at 0 */;
            
            #line 87 "../../../toolkit/www/examples/xs/xs.m"
             sprintf(buf, "XR+%08x", adr);

            
            
            
          }
          
          break;
        case 3: 
          { 
            unsigned adr = (MATCH_w_16_0 & 0x3ff) /* adr at 0 */;
            
            #line 88 "../../../toolkit/www/examples/xs/xs.m"
             sprintf(buf, "XR-%08x", adr);

            
            
            
          }
          
          break;
        default: assert(0);
      } /* (MATCH_w_16_0 >> 10 & 0x3) -- md at 0 --*/ 
    
  }goto MATCH_finished_e; 
  
  MATCH_finished_e: (void)0; /*placeholder for label*/
  
}

#line 91 "../../../toolkit/www/examples/xs/xs.m"
  return buf;
}
#line 218 "xs.nw"
static unsigned getword(FILE *fp);
void load_and_goXS(Machine m, FILE *loadfile) {
  unsigned LA, N, START;  
  unsigned i;
  LA = getword(loadfile);
  N  = getword(loadfile);
  for (i = 0; i < N; i++)
    m->MEM[validate(LA+i)] = getword(loadfile);
  START = getword(loadfile);
  interpretXS(m, START);
}
#line 231 "xs.nw"
static unsigned getword(FILE *fp) {
  unsigned u;
  u = 0;
  u |= ((unsigned char) getc(fp)) <<  0;
  u |= ((unsigned char) getc(fp)) <<  8;
  u |= ((unsigned char) getc(fp)) << 16;
  u |= ((unsigned char) getc(fp)) << 24;
  return u;
}

#line 253 "xs.nw"
main(int argc, char *argv[]) {
  FILE *loadfile;
  Machine m;
  
#line 265 "xs.nw"
if (argc != 2) {
  fprintf(stderr, "Usage: %s loadfile\n", argv[0]);
  exit(1);
}
#line 257 "xs.nw"
  loadfile = fopen(argv[1], "r");
  
#line 270 "xs.nw"
if (loadfile == NULL) {
  fprintf(stderr, "%s: Could not open file %s for read\n", argv[0], argv[1]);
  exit(1);
}
#line 259 "xs.nw"
  m = (Machine) malloc(sizeof(*m));
  assert(m);
  load_and_go(m, loadfile);
}


